### PR TITLE
✨:sparkles: Check Unique Ids

### LINKED ISSUE #163: ✨Check if IDs are Unique
#### Describe your issue

The new Property Panel doesn't check if the ID's are unique. 

#### Possible solution

Use aurelia validation to check if the id's are unique to all other elements in the diagramm.
Basics for that are already implement in the basics section.

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by MiguelACarballo**
## What did you change?

This PR adds a feature to check whether an ID or Process-ID is unique.
If not, you cannot safe and a warning appears.

Fixes #163 

## How can others test the changes?

1. Checkout the branch.
2. npm install
3. au run --watch
4. Open a diagram, open an element and type an existing Id
5. A warning should appear inside the ID field.

Same for Process ID.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.)
- [x] I've merged the `develop` branch into my branch before finishing this PR.
- [x] I've **not added any other changes** than the ones described above.
- [x] I've mentioned all **PRs, which relate to this one**

### CODE REVIEW

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts [hunk 1]
```diff
@@ -119,30 +99,43 @@ export class BasicsSection implements ISection {
...
-    for (const result of event.results) {
-      if (result.rule.property.displayName !== 'elementId') {
-        continue;
-      }
-      if (result.valid === false) {
-        this.validationError = true;
-        document.getElementById(result.rule.property.displayName).style.border = '2px solid red';
-      } else {
-        document.getElementById(result.rule.property.displayName).style.border = '';
+  private checkId(): void {
```
Comment (Vyperus) [inline src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: This method does something more than "checking" an Id. Please consider using a more meaningful name or even better split this method in more smaller and more readable methods.

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts [hunk 2]
```diff
@@ -119,30 +99,43 @@ export class BasicsSection implements ISection {
...
+        if (element.type !== 'label') {
+          return element.businessObject.id === this.businessObjInPanel.id;
+        }
       }
+      return false;
+    });
+
+    const hasSameIdWithOtherElements: Boolean = elementsWithSameId.length > 0;
+    const idIsEmpty: Boolean = this.businessObjInPanel.id.length <= 0;
+    if (hasSameIdWithOtherElements) {
```
Comment (Vyperus) [inline src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: There is an "aurelia-way" of doing this whole "pushing to errors array"-thing. @MiguelACarballo already used it in another method. Also consider: http://aurelia.io/docs/plugins/validation#introduction

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 3]
```diff
@@ -94,4 +90,23 @@ export class PoolSection implements ISection {
     }
   }
 
+  private checkId(): void {
```
Comment (Vyperus) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: This mehtod does much more than "checking" an Id, please use more meaningful names or, even better, split it into smaller, more readable, methods.

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts [hunk 4]
```diff
@@ -145,4 +142,30 @@ export class BasicsSection implements ISection {
...
+          return element.businessObject.id === this.businessObjInPanel.id;
+        }
+      }
+      return false;
+    });
+
+    return elementsWithSameId.length === 0;
+  }
+
+  private checkIsIdUnique(): void {
```
Comment (Vyperus) [inline src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: This method is not "checking" if an id is unique. It sets the ValidationRules. With these the ValidationController validates the object. The rules here are saying that the id can't be empty AND has to be unique. My suggestion: 
`private setValidationRules(): void {...}`

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts [hunk 5]
```diff
@@ -20,12 +21,12 @@ export class BasicsSection implements ISection {
   private modeler: IBpmnModeler;
   private moddle: IBpmnModdle;
   private elementInPanel: IShape;
-  private saveId: string;
+  private previousId: string;
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: Ich würde mit zwei variablen arbeiten:

`currentId` & `fallbackID`.

Ein `save` lößt das Update auf den der `currentId` aus; in pseudocode so:

```ts

save() {
  fallbackId = currentId;
  currentId = newId;

  if (idIsNotUnique(currentId))
    currentId = fallbackId;
  else
    fallbackId = currentId;
}
```

Comment (cmg-dev) [reply src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: passt das?

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts [hunk 6]
```diff
@@ -34,28 +35,24 @@ export class BasicsSection implements ISection {
...
     this.businessObjInPanel = model.elementInPanel.businessObject;
-    this.saveId = model.elementInPanel.businessObject.id;
+    this.previousId = model.elementInPanel.businessObject.id;
     this.modeling = model.modeler.get('modeling');
     this.moddle = model.modeler.get('moddle');
     this.modeler = model.modeler;
 
     this.validationController.subscribe((event: ValidateEvent) => {
-      this.validateForm(event);
+      this.validateId(event);
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: `validateFormId` das ist spezifischer und richtiger.

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts [hunk 7]
```diff
@@ -145,4 +142,30 @@ export class BasicsSection implements ISection {
       }
     }
   }
+
+  private isIdUnique(id: string): boolean {
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: Hier könnte man den Namen noch ändern

`_idIsUnique()`

Der Vorteil wäre: Es ließt sich im if besser.

```ts
if (idIsUnique(currentId))
```

Ist aber eine kann-muss-aber-nicht-Regel.

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 8]
```diff
@@ -69,13 +70,14 @@ export class PoolSection implements ISection {
...
     this.businessObjInPanel.processRef.id = '';
+    this.validationController.validate();
   }
 
   private clearName(): void {
     this.businessObjInPanel.processRef.name = '';
   }
 
-  private validateForm(event: ValidateEvent): void {
+  private validateId(event: ValidateEvent): void {
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: `validateFormId`

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 9]
```diff
@@ -94,4 +96,39 @@ export class PoolSection implements ISection {
     }
   }
 
+  private isIdUnique(id: string): boolean {
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: s.o. _formIdIsUnique() wäre eine Möglichkeit.

Bitte den Underscore beachten.

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 10]
```diff
@@ -94,4 +96,39 @@ export class PoolSection implements ISection {
...
+      return rootElement.id;
+    });
+
+    const currentId: number = elementIds.indexOf(this.businessObjInPanel.processRef.id);
+    elementIds.splice(currentId, 1);
+
+    return !elementIds.includes(id);
+  }
+
+  private setValidationRules(): void {
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: Alle Privaten Methoden mit `_`-Prefix.

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 11]
```diff
@@ -94,4 +96,39 @@ export class PoolSection implements ISection {
...
+    ValidationRules.ensure((businessObject: IModdleElement) => businessObject.id)
+    .displayName('processId')
+    .required()
+      .withMessage(`Process-Id cannot be blank.`)
+    .then()
+    .satisfies((id: string) => this.isIdUnique(id) && this.isProcessIdUnique(id))
+      .withMessage(`Process-Id already exists.`)
+    .on(this.businessObjInPanel.processRef);
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: Einrückung passt nicht!

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts (hunk 10 above)
Comment (cmg-dev) [reply src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: Hier gab es mal eine Diskussion dazu: 

https://github.com/5minds/typescript-guidelines/issues/10

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 12]
```diff
@@ -19,36 +22,34 @@ export class PoolSection implements ISection {
...
-      .required()
-      .withMessage(`Process-Id cannot be blank.`)
-      .on(this.businessObjInPanel.processRef || {});
+    this.setValidationRules();
   }
 
   public detached(): void {
     if (this.validationError) {
-      this.businessObjInPanel.processRef.id = this.saveProcessId;
+      this.businessObjInPanel.processRef.id = this.previousId;
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: previousProcessRefId, wenn.

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 13]
```diff
@@ -19,36 +22,34 @@ export class PoolSection implements ISection {
...
+  private previousId: string;
 
   constructor(controller?: ValidationController) {
     this.validationController = controller;
   }
 
   public activate(model: IPageModel): void {
     if (this.validationError) {
-      this.businessObjInPanel.processRef.id = this.saveProcessId;
+      this.businessObjInPanel.processRef.id = this.previousId;
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: previousProcessRefId, wenn.

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 14]
```diff
@@ -19,36 +22,34 @@ export class PoolSection implements ISection {
...
       this.validationController.validate();
     }
 
     this.businessObjInPanel = model.elementInPanel.businessObject;
-    this.saveProcessId = this.businessObjInPanel.processRef.id;
+    this.previousId = this.businessObjInPanel.processRef.id;
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: previousProcessRefId, wenn.

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 15]
```diff
@@ -19,36 +22,34 @@ export class PoolSection implements ISection {
 
   private businessObjInPanel: IPoolElement;
   private modeler: IBpmnModeler;
-  private saveProcessId: string;
+  private moddle: IBpmnModdle;
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: bpmnModdle fände ich schon einen besseren Namen, aber auch nicht viel besser.

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 16]
```diff
@@ -19,36 +22,34 @@ export class PoolSection implements ISection {
...
-      this.businessObjInPanel.processRef.id = this.saveProcessId;
+      this.businessObjInPanel.processRef.id = this.previousId;
       this.validationController.validate();
     }
 
     this.businessObjInPanel = model.elementInPanel.businessObject;
-    this.saveProcessId = this.businessObjInPanel.processRef.id;
+    this.previousId = this.businessObjInPanel.processRef.id;
+    this.moddle = model.modeler.get('moddle');
     this.modeler = model.modeler;
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: Reihenfolge tauschen.

erst `Modeler` holen, dann `moddle`.

```ts
this.modeler = model.modeler;
this.bpmnModdle = this.modeler.get('moddle');
```

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 17]
```diff
@@ -19,36 +22,34 @@ export class PoolSection implements ISection {
...
 
     this.validationController.subscribe((event: ValidateEvent) => {
-      this.validateForm(event);
+      this.validateId(event);
```
Comment (cmg-dev) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: validateFormId

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts (hunk 5 above)
Comment (cmg-dev) [reply src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: Noch eine Sache: https://github.com/process-engine/bpmn-studio/pull/178/files#diff-ae363f44f43dcd14cc56639982a06869R10

sowas bricht echt schnell auseinander. Bitte besser machen.

File: src/modules/property-panel/indextabs/forms/sections/basics/basics.ts [hunk 18]
```diff
@@ -91,19 +155,28 @@ export class BasicsSection implements ISection {
     }
   }
 
-  private updateId(): void {
+  private _updateId(): void {
+    this.validationController.validate();
+    if (this.validationController.errors.length > 0) {
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/forms/sections/basics/basics.ts]: please move this condition to a context giving variable

File: src/modules/property-panel/indextabs/forms/sections/basics/basics.ts [hunk 19]
```diff
@@ -91,19 +155,28 @@ export class BasicsSection implements ISection {
...
+      this._resetId();
+    }
+
+    if (this.selectedForm === null || this.selectedForm.id === '') {
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/forms/sections/basics/basics.ts]: please move this condition to a context giving variable

File: src/modules/property-panel/indextabs/forms/sections/basics/basics.ts [hunk 20]
```diff
@@ -198,29 +274,123 @@ export class BasicsSection implements ISection {
...
+        this.validationError = true;
+        document.getElementById(result.rule.property.displayName).style.border = '2px solid red';
+      } else {
+        document.getElementById(result.rule.property.displayName).style.border = '2px solid green';
+      }
+    }
+  }
+
+  private _hasFormSameIdAsSelected(forms: Array<IForm>): boolean {
+    for (const form of forms) {
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/forms/sections/basics/basics.ts]: you can improve the readability if this function like this:
```TypeScript
private _hasFormSameIdAsSelected(forms: Array<IForm>): boolean {

  const unselectedFormWithSameId: IForm = forms.find((form: IForm) => {

    const formHasSameIdAsSelectedForm: boolean = form.id === this.selectedForm.id;
    const formIsNotSelectedForm: boolean = form !== this.selectedForm;

    return formHasSameIdAsSelectedForm && formIsNotSelectedForm;
  });

  return unselectedFormWithSameId !== undefined;
}
```

File: src/modules/property-panel/indextabs/forms/sections/basics/basics.ts [hunk 21]
```diff
@@ -198,29 +274,123 @@ export class BasicsSection implements ISection {
...
+
+  private _getFormDataFromBusinessObject(businessObject: IModdleElement): IFormElement {
+    const extensionElement: IExtensionElement = businessObject.extensionElements;
+    const hasNoExtensionElements: boolean = extensionElement === undefined;
+    if (hasNoExtensionElements) {
+      return;
+    }
+
+    const extensions: Array<IModdleElement> = extensionElement.values;
+    for (const extension of extensions) {
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/forms/sections/basics/basics.ts]: You can improve the readability by replacing this loop with an `Array.find`:

```TypeScript
return extensions.find((extension: IModdleElement) => {
  const isFormData: boolean = extension.$type === 'camunda:FormData';

  return isFormData;
});

File: src/modules/property-panel/indextabs/forms/sections/basics/basics.ts [hunk 22]
```diff
@@ -198,29 +274,123 @@ export class BasicsSection implements ISection {
...
+    }
+
+    return;
+  }
+
+  private _getFormsById(id: string): Array<IShape> {
+    const elementRegistry: IElementRegistry = this.modeler.get('elementRegistry');
+
+    const formsWithId: Array<IShape> = elementRegistry.filter((element: IShape) => {
+      const currentBusinessObj: IModdleElement = element.businessObject;
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/forms/sections/basics/basics.ts]: We don't need to shorten variable names, please use `currentBusinessObject `

File: src/modules/property-panel/indextabs/forms/sections/basics/basics.ts [hunk 23]
```diff
@@ -198,29 +274,123 @@ export class BasicsSection implements ISection {
...
+
+    return isIdUnique;
+  }
+
+  private _setValidationRules(): void {
+    ValidationRules
+      .ensure((form: IForm) => form.id)
+      .displayName('formId')
+      .required()
+      .withMessage(`Id cannot be blank.`)
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/forms/sections/basics/basics.ts]: You don't need to use a template-string here, a regular string will do.

File: src/modules/property-panel/indextabs/forms/sections/basics/basics.ts [hunk 24]
```diff
@@ -198,29 +274,123 @@ export class BasicsSection implements ISection {
...
+      .then()
+      .satisfies((id: string) => this._formIdIsUnique(id))
+      .withMessage(`Id already exists.`)
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/forms/sections/basics/basics.ts]: You don't need to use a template-string here, a regular string will do.

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts [hunk 25]
```diff
@@ -145,4 +152,30 @@ export class BasicsSection implements ISection {
       }
     }
   }
+
+  private _formIdIsUnique(id: string): boolean {
+    const elementRegistry: IElementRegistry = this.modeler.get('elementRegistry');
+    const elementsWithSameId: Array<IShape> =  elementRegistry.filter((element: IShape) => {
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: there are two spaces after the `=`

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts [hunk 26]
```diff
@@ -145,4 +152,30 @@ export class BasicsSection implements ISection {
...
+      if (element.businessObject !== this.businessObjInPanel) {
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: you can improve readability here by moving the conditions to variables:

```TypeScript
const elementIsBusinessObjectInPanel: boolean = element.businessObject === this.businessObjInPanel;
if (elementIsBusinessObjectInPanel) {
  return false;
}

const elementIsOfTypeLabel: boolean = element.type === 'label';
if (elementIsOfTypeLabel) {
  return false;
}

const elementHasSameId: boolean = element.businessObject.id === this.businessObjInPanel.id;

return elementHasSameId;

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts [hunk 27]
```diff
@@ -145,4 +152,30 @@ export class BasicsSection implements ISection {
...
+
+    return elementsWithSameId.length === 0;
+  }
+
+  private _setValidationRules(): void {
+    ValidationRules
+      .ensure((businessObject: IModdleElement) => businessObject.id)
+      .displayName('elementId')
+      .required()
+        .withMessage(`Id cannot be blank.`)
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: You don't need to use a template-string here, a regular string will do.

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts [hunk 28]
```diff
@@ -145,4 +152,30 @@ export class BasicsSection implements ISection {
...
+      .then()
+      .satisfies((id: string) => this._formIdIsUnique(id))
+        .withMessage(`Id already exists.`)
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: You don't need to use a template-string here, a regular string will do.

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 29]
```diff
@@ -1,7 +1,10 @@
 import {inject} from 'aurelia-framework';
 import {ValidateEvent, ValidationController, ValidationRules} from 'aurelia-validation';
+import { Config } from 'protractor';
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: remove the spaces within the curly braces

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 30]
```diff
@@ -94,4 +97,39 @@ export class PoolSection implements ISection {
     }
   }
 
+  private _formIdIsUnique(id: string): boolean {
+    const elementRegistry: IElementRegistry = this.modeler.get('elementRegistry');
+    const elementsWithSameId: Array<IShape> =  elementRegistry.filter((element: IShape) => {
+      if (element.businessObject !== this.businessObjInPanel) {
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: Like above, you can improve readability here by moving the conditions to variables:

```TypeScript
const elementIsBusinessObjectInPanel: boolean = element.businessObject === this.businessObjInPanel;
if (elementIsBusinessObjectInPanel) {
  return false;
}

const elementIsOfTypeLabel: boolean = element.type === 'label';
if (elementIsOfTypeLabel) {
  return false;
}

const elementHasSameId: boolean = element.businessObject.id === this.businessObjInPanel.id;

return elementHasSameId;

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 31]
```diff
@@ -94,4 +97,39 @@ export class PoolSection implements ISection {
...
+    elementIds.splice(currentId, 1);
+
+    return !elementIds.includes(id);
+  }
+
+  private _setValidationRules(): void {
+    ValidationRules.ensure((businessObject: IModdleElement) => businessObject.id)
+    .displayName('processId')
+    .required()
+    .withMessage(`Process-Id cannot be blank.`)
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: You don't need to use a template-string here, a regular string will do.

File: src/modules/property-panel/indextabs/general/sections/pool/pool.ts [hunk 32]
```diff
@@ -94,4 +97,39 @@ export class PoolSection implements ISection {
...
+    .then()
+    .satisfies((id: string) => this._formIdIsUnique(id) && this._isProcessIdUnique(id))
+    .withMessage(`Process-Id already exists.`)
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/general/sections/pool/pool.ts]: You don't need to use a template-string here, a regular string will do.

File: src/modules/property-panel/indextabs/general/sections/basics/basics.ts [hunk 33]
```diff
@@ -145,4 +152,46 @@ export class BasicsSection implements ISection {
       }
     }
   }
+
+  private _formIdIsUnique(id: string): boolean {
+    const elementRegistry: IElementRegistry = this.modeler.get('elementRegistry');
+
+    const elementsWithSameId: Array<IShape> =  elementRegistry.filter((element: IShape) => {
```
Comment (heikomat) [inline src/modules/property-panel/indextabs/general/sections/basics/basics.ts]: There are still two spaces after the `=`

**Pull request merged**